File: frontend/src/utils/regionScope.ts
// Canonical persisted marker meaning "force this channel unscoped / plain flood"
// (mirrors the backend UNSCOPED_OVERRIDE_MARKER). Distinct from null, which means
// "inherit the global scope".
export const UNSCOPED_OVERRIDE_MARKER = '*';

export function isUnscopedMarker(scope: string | null | undefined): boolean {
  return scope === UNSCOPED_OVERRIDE_MARKER;
}

export function stripRegionScopePrefix(scope: string | null | undefined): string {
  if (!scope) return '';
  return scope.startsWith('#') ? scope.slice(1) : scope;
}

// Compact, human-readable label for a channel's persisted flood-scope override,
// suitable for a header badge. Returns null when there is no override (inherit),
// "unscoped" for the unscoped marker, and the region name as-is otherwise.
export function floodScopeOverrideLabel(scope: string | null | undefined): string | null {
  if (!scope) return null;
  if (isUnscopedMarker(scope)) return 'unscoped';
  return scope;
}

File: app/region_scope.py
"""Helpers for normalizing MeshCore flood-scope / region names."""

# Canonical persisted marker for "force unscoped/plain flood". Stored verbatim in
# the per-channel ``flood_scope_override`` column to mean "this channel is unscoped
# even if a global region is set" — distinct from NULL, which means "inherit global".
UNSCOPED_OVERRIDE_MARKER = "*"

# All values that denote explicit unscoped/plain flood, matching firmware parity.
_UNSCOPED_SENTINELS = {"", "0", UNSCOPED_OVERRIDE_MARKER}


def is_unscoped(scope: str | None) -> bool:
    """True if ``scope`` denotes an explicit unscoped/plain-flood request.

    Note: an empty string counts as unscoped here. Callers that need to treat
    blank as "no opinion / inherit" (e.g. the channel-override API) must check for
    blank *before* calling this.
    """
    return (scope or "").strip() in _UNSCOPED_SENTINELS


def normalize_region_scope(scope: str | None) -> str:
    """Normalize a user-facing region scope into MeshCore's internal form.

    Region names are user-facing plain strings like ``Esperance``. Internally,
    MeshCore still expects hashtag-style names like ``#Esperance``.

    Backward compatibility / firmware parity:
    - blank/None stays unscoped (``""``)
    - ``"0"`` and ``"*"`` also mean explicit unscoped/plain flood
    - existing leading ``#`` is preserved
    """

    stripped = (scope or "").strip()
    if stripped in _UNSCOPED_SENTINELS:
        return ""
    if stripped.startswith("#"):
        return stripped
    return f"#{stripped}"
